### DISCUSSION TITLE
Alert customConstraints with new field and migrate from other table

### DISCUSSION THREAD

**Discussion by d3xt3r2909** [simolus3/drift#1432] [General]
Hi,

I have a little bit advanced situation and would love to figure out what would be the optimal solution for migration. 

So I have 3 tables: `User`, `UserCategory`, `Category`. 

Columns per table: 

`User` -> `Username`, `UserId`, ...
`Category` -> `CategoryId` ...

And .. 
```

@DataClassName('UserCategory')
class UserCategory extends Table {
  TextColumn get username => text()();

  TextColumn get categoryId => text()();

  @override
  List<String> get customConstraints => ['UNIQUE (username, category_id)'];
}

```
-----------

Now, because I have a requirement to make `username` optional, I will need to replace `username` inside `UserCategory` table with userId. 

This is how table should look after migration.. 

```
@DataClassName('UserCategory')
class UserCategory extends Table {
  TextColumn get userId => text()();

  TextColumn get categoryId => text()();

  @override
  List<String> get customConstraints => ['UNIQUE (user_id, category_id)'];
}
```

-----------

One of the ideas would be to use select on `UserCategory`, then to update columns, but is there anything better?

-----------

Really thank you for this plugin, I would say that's one of the best-written plugins on pub.dev

**Comment by simolus3** [comment]
I think you can do these transformations by using a `TableMigration` and a subquery in a [complex migration](https://moor.simonbinder.eu/docs/advanced-features/migrations/#complex-migrations). I haven't tried this, but I hope something like this works in a migration callback:

```dart
// old "username" column on the "UserCategory" table that we're about to replace
final oldName = GeneratedColumn<String>('username', 'user_category', false); 
// Use a subquery to map user names from UserCategory to their matching ids
final userIdForName = subqueryExpression<int>(selectOnly(users)
    ..addColumns([users.id])
    ..where(users.name.equalsExp(oldName)),
);

await m.alterTable(
  TableMigration(
    userCategory,
    newColumns: [userCategory.userId],
    columnTransformer: {
      // replace the name with the matching id
      userCategory.userId: userIdForName,
    }
  ),
);
```

**Reply by d3xt3r2909** [reply]
Hey, good idea, and thanks for the reply. I Will try it soon and post results

**Reply by d3xt3r2909** [reply]
That's it, with small modificiations. Thanks!
